src/main/scala/se/lth/immun/TargetDeisotoper.scala

```diff
@@ -1,57 +1,162 @@
 package se.lth.immun
 
 import scala.collection.mutable.ArrayBuffer
+import scala.collection.mutable.Queue
+import scala.collection.mutable.HashSet
+
+import akka.actor._
 
 import se.lth.immun.chem.Peptide
 import se.lth.immun.chem.Constants
 import se.lth.immun.chem.IsotopeDistribution
 
-class TargetDeisotoper(val params:DinosaurParams) extends Timeable {
+class TargetDeisotoper(val params:DinosaurParams) extends Actor with Timeable {
 
 	import Cluster._
 	
 	implicit val p = params
 	
-	def deisotope(
-			hills:Array[Hill], 
-			targets:Seq[Target], 
-			specTime:Seq[Double]
-	):(Seq[Seq[Cluster.Edge]], Seq[IsotopePattern]) = {
+	val toProcess 			= new Queue[(Int, Seq[Target])]
+	val beingProcessed 		= new HashSet[Int]
+	val completedPatterns	= new ArrayBuffer[Seq[IsotopePattern]]
+	
+	var hillsByMz:Array[Hill] = _
+	var hillsMzs:Array[Double] = _
+	var hillsRts:Array[Double] = _
+	var hillsScanIndices:Array[(Int, Int)] = _
+	var specTime:Seq[Double] = _
+	var customer:ActorRef = _
+	
+	val targetBatchSize = 1000
+	
+	def receive = {
+		case TargetDeisotope(hs, targets, st) => 
+			timer.start
+			
+			/*
+			 * Using a list for the hills is extremely slow for many hills
+			 * an array is asympotically faster
+			 */
+			val hills = hs.toArray
+			specTime = st
+			customer = sender
+			
+			println("=== IN TARGETED MODE - BEWARE!!! ===")
+		  hillsByMz = hills.sortBy(_.total.centerMz)
+		  hillsMzs = hillsByMz.map(h => h.total.centerMz)
+		  hillsRts = hillsByMz.map(h => h.accurateApexRt(specTime))
+		  hillsScanIndices = hillsByMz.map(h => (h.scanIndex.head, h.scanIndex.last))
+		  
+		  if (params.verbose)
+			  println("deisotoping based on targets...")
+			
+      val batchTargetList = targets.sortBy(_.mz).grouped(targetBatchSize).toList
+      for (i <- 0 until batchTargetList.length)
+			  toProcess += i -> batchTargetList(i)
```

Comment:
Here you do random access on a list (`batchTargetList`). `zipWithIndices` or using an array would be better.

Reply:
Agreed, though the impact is probably minimal.